.github/workflows/main.yaml

```
@@ -0,0 +1,127 @@
+name: CICD-Workflow
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+
+env:
+  IMAGE_NAME: "nodeenergy.azurecr.io/ppa-predictions:${{ github.sha }}"
+  CONTEXT: "optinode-staging"
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Build and export
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          tags: ${{ env.IMAGE_NAME }}
+          outputs: type=docker,dest=/tmp/testimage.tar
+      - name: upload image artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: testimage
+          path: /tmp/testimage.tar
+
+  test:
+    runs-on: ubuntu-latest
+    needs: [ build ]
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Download artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: testimage
+          path: /tmp
+      - name: Load image
+        run: docker load --input /tmp/testimage.tar
+
+      - name: Run pytest in Docker container
+        run: docker run -v ./tests/:/code/tests --rm ${{ env.IMAGE_NAME }} pytest
+
+  publish_image:
+    if: ${{ github.ref == 'refs/heads/main' }}
```

IMO this would be better as 
```suggestion
    if: github.event_name == 'push'
```